src/ucai/core/utils/callable_utils.py

```diff
@@ -0,0 +1,142 @@
+import ast
+import inspect
+from textwrap import dedent
+from typing import Any, Callable, get_type_hints
+
+from ucai.core.utils.type_utils import python_type_to_sql_type, validate_container_type
+
+FORBIDDEN_PARAMS = ['self', 'cls']
+
+def extract_function_body(func: Callable[..., Any]) -> str:
+    """Extracts the body of a function as a string without the signature or docstring, preserving indentation."""
+    parsed_source, source_lines = parse_dedented_source(func)
+    
+    function_def = parsed_source.body[0]
+
+    body_start_line = function_def.body[0].lineno - 1
+    
+    if isinstance(function_def.body[0], ast.Expr) and isinstance(function_def.body[0].value, ast.Constant):
+        body_start_line += len(function_def.body[0].value.s.splitlines())
+
+    function_body_lines = source_lines[body_start_line:]
+
+    min_indent = min((len(line) - len(line.lstrip()) for line in function_body_lines if line.strip()), default=0)
+    
+    function_body = "\n".join([line[min_indent:] if len(line) >= min_indent else line for line in function_body_lines])
+
+    return function_body
+
+def parse_docstring(docstring: str) -> dict:
```

Comment:
```suggestion
def parse_docstring(docstring: str) -> dict[???, ???]:
```

what's the key and value type?

Reply:
ahhhh ironic that I'm rejecting these undefined collections and I forget to supply them. Thanks